### ISSUE TITLE
TDISP keywords missing when writing Table

### ISSUE THREAD

**Issue by weiaip** [astropy/astropy#6806]
When writing a astropy.table.Table to disk in FITS format, any column formatting is lost while it should get written to TDISP keywords.

Example code:
```python
from astropy.table import Table
a = [1, 4, 5]
b = [2.0, 5.0, 8.2]
c = ['x', 'y', 'z']
t = Table([a, b, c], names=('a', 'b', 'c'), meta={'name': 'first table'})
t['a'].format = '%3d'
t['b'].format = '%4.2f'
t.write('tabletest.fits')
```

Results in table FITS header:
```
XTENSION= 'BINTABLE'           / binary table extension
BITPIX  =                    8 / array data type
NAXIS   =                    2 / number of array dimensions
NAXIS1  =                   17 / length of dimension 1
NAXIS2  =                    3 / length of dimension 2
PCOUNT  =                    0 / number of group parameters
GCOUNT  =                    1 / number of groups
TFIELDS =                    3 / number of table fields
TTYPE1  = 'a       '
TFORM1  = 'K       '
TTYPE2  = 'b       '
TFORM2  = 'D       '
TTYPE3  = 'c       '
TFORM3  = '1A      '
NAME    = 'first table'
END
```

**Comment by taldcroft** [comment]
:+1: in general.

Your example shows `TFORM` but your words say `TDISP`.  My understanding is that `TFORM` is about how the value is *encoded* in the FITS file, so we want `TDISP`.  See below, which I include for myself and any other non-FITS-experts.

One complication will be converting between any allowed Python format and the required Fortran format specification.

```
KEYWORD:   TDISPn
REFERENCE: FITS Standard 
STATUS:    reserved
HDU:       table
VALUE:     string
COMMENT:   display format
DEFINITION: The value field of this indexed keyword shall contain a
character string describing the format recommended for the display of
the contents of field n.  If the table value has been scaled, the
physical value shall be displayed.  All elements in a field shall be
displayed with a single, repeated format. For purposes of display, each
byte of bit (type X) and byte (type B) arrays is treated as an unsigned
integer. Arrays of type A may be terminated with a zero byte.  Only the
format codes in Table 8.6, discussed in section 8.3.4 of the FITS
Standard, are permitted for encoding. The format codes must be
specified in upper case.  If the Bw.m, Ow.m, and Zw.m formats are not
readily available to the reader, the Iw.m display format may be used
instead, and if the ENw.d and ESw.d formats are not available, Ew.d may
be used.  The meaning of this keyword is not defined for fields of type
P in the Standard but may be defined in conventions using such fields.

KEYWORD:   TFORMn
REFERENCE: FITS Standard 
STATUS:    mandatory
HDU:       table
VALUE:     string
COMMENT:   column data format
DEFINITION: The value field of this indexed keyword shall contain a
character string describing the format in which field n is encoded in a
'TABLE' or 'BINTABLE' extension.
```

**Comment by weiaip** [comment]
The example above is just showing the bug, i.e. the current state without TDISP keywords. Since TFORM is mandatory, and as you say, determines the binary encoding, that cannot be touched.

**Comment by SaraOgaz** [comment]
So I'm planning on fixing this, as we need it to roundtrip some of our FITS Binary tables in some reference files.  I'll probably have some questions when I get around to it.

**Comment by taldcroft** [comment]
@SaraOgaz - #7147 will give you round-tripping.  Is this good enough or do you actually need `TDISP`?

**Comment by SaraOgaz** [comment]
Will check, thanks @taldcroft!

**Comment by SaraOgaz** [comment]
Hmm, I grabbed the current master (3.1.dev21371) and it's still getting left out.  There are actually a few other small issues that have popped up, but I was planning on waiting until #7145 was resolved, and go from there. I can file some new issues for this stuff since at this point I think it's unrelated to this particular issue.

**Comment by SaraOgaz** [comment]
Update again... look like my problem might be on the other end (reading in my binary table), will test again after that's figured out.

**Comment by taldcroft** [comment]
@SaraOgaz - here's what I see in current master:
```
In [1]: astro
astropy=3.1.dev21450

In [2]: t = simple_table()

In [3]: t
Out[3]: 
<Table length=3>
  a      b     c  
int64 float64 str1
----- ------- ----
    1     1.0    c
    2     2.0    d
    3     3.0    e

In [4]: t['b'].format = '8.4f'

In [5]: t
Out[5]: 
<Table length=3>
  a      b      c  
int64 float64  str1
----- -------- ----
    1   1.0000    c
    2   2.0000    d
    3   3.0000    e

In [6]: t.write('junk.fits', overwrite=True)

In [7]: t2 = Table.read('junk.fits')

In [8]: t2
Out[8]: 
<Table length=3>
  a      b       c   
int64 float64  bytes1
----- -------- ------
    1   1.0000      c
    2   2.0000      d
    3   3.0000      e

In [9]: !fitsheader junk.fits
# HDU 0 in junk.fits:
SIMPLE  =                    T / conforms to FITS standard                      
BITPIX  =                    8 / array data type                                
NAXIS   =                    0 / number of array dimensions                     
EXTEND  =                    T                                                  

# HDU 1 in junk.fits:
XTENSION= 'BINTABLE'           / binary table extension                         
BITPIX  =                    8 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                   17 / length of dimension 1                          
NAXIS2  =                    3 / length of dimension 2                          
PCOUNT  =                    0 / number of group parameters                     
GCOUNT  =                    1 / number of groups                               
TFIELDS =                    3 / number of table fields                         
TTYPE1  = 'a       '                                                            
TFORM1  = 'K       '                                                            
TTYPE2  = 'b       '                                                            
TFORM2  = 'D       '                                                            
TTYPE3  = 'c       '                                                            
TFORM3  = '1A      '                                                            
COMMENT --BEGIN-ASTROPY-SERIALIZED-COLUMNS--                                    
COMMENT datatype:                                                               
COMMENT - {name: a, datatype: int64}                                            
COMMENT - {name: b, datatype: float64, format: 8.4f}                            
COMMENT - {name: c, datatype: string}                                           
COMMENT meta:                                                                   
COMMENT   __serialized_columns__: {}                                            
COMMENT --END-ASTROPY-SERIALIZED-COLUMNS--                                      
```

**Comment by SaraOgaz** [comment]
Yeah @taldcroft, I suspect this end of things is working fine with your fix (thanks!), it's the reading in of the FITS file to the table that isn't saving out the TDSIP keywords to column attributes:  https://github.com/astropy/astropy/issues/7165

**Comment by SaraOgaz** [comment]
@taldcroft @saimn actually wait a second... is the TDISPn keyword suppose to go to and from the column.format attribute?  #6806 makes it look like column.format maps to TDISP, but in @taldcroft's above example it maps to a COMMENT keyword...

**Comment by saimn** [comment]
Fixed by #7226.
